File: 063. Unique Paths II.py
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        m = len(obstacleGrid)
        if m == 0:
            return 0
        n = len(obstacleGrid[0])
        matrix = obstacleGrid[:]
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 1:
                    matrix[i][j] = 0
                else:
                    matrix[i][j] = 1
        for i in range(1, m):
            if matrix[i][0] != 0:
                matrix[i][0] = matrix[i - 1][0]
        for j in range(1, n):
            if matrix[0][j] != 0:
                matrix[0][j] = matrix[0][j - 1]
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][j] != 0:
                    matrix[i][j] = matrix[i - 1][j] + matrix[i][j - 1]
        return matrix[m - 1][n - 1]

File: 063. Unique Paths II.swift
class Solution {
    func uniquePathsWithObstacles(_ obstacleGrid: [[Int]]) -> Int {
        let m = obstacleGrid.count
        let n = obstacleGrid[0].count
        
        var map = Array(repeating: Array(repeating: 0, count: n), count: m)
        
        for i in 0 ..< m {
            if obstacleGrid[i][0] == 1 {
                break
            }
            map[i][0] = 1
        }
        
        for j in 0 ..< n {
            if obstacleGrid[0][j] == 1 {
                break
            }
            map[0][j] = 1
        }
        
        for i in 1 ..< m {
            for j in 1 ..< n {
                if obstacleGrid[i][j] == 1  {
                    map[i][j] = 0
                    continue
                }
                map[i][j] = map[i-1][j] + map[i][j-1]
            }
        }
        
        return map[m-1][n-1]
    }
}
